src/main/java/com/iota/iri/service/garbagecollector/GarbageCollector.java

```diff
@@ -0,0 +1,323 @@
+
+    /**
+     * List of cleanup jobs that shall get processed by the garbage collector (grouped by their class).
+     */
+    private final HashMap<Class<? extends GarbageCollectorJob>, ArrayDeque<GarbageCollectorJob>> garbageCollectorJobs = new HashMap<>();
+
+    /**
+     * The constructor of this class stores the passed in parameters for future use and restores the previous state of
+     * the garbage collector if there is a valid one (to continue with cleaning up after IRI restarts).
+     *
+     * Before restoring the garbage collector state it registers the available job types, so we know how to parse and
+     * process the found jobs.
+     */
+    public GarbageCollector(Tangle tangle, SnapshotManager snapshotManager, TipsViewModel tipsViewModel) {
+        this.tangle = tangle;
+        this.snapshotManager = snapshotManager;
+        this.tipsViewModel = tipsViewModel;
+
+        MilestonePrunerJob.registerInGarbageCollector(this);
+        UnconfirmedSubtanglePrunerJob.registerInGarbageCollector(this);
+
+        restoreCleanupJobs();
+    }
+
+    /**
+     * This method adds a job to the GarbageCollector, that will get processed on its next run.
+     *
+     * It first registers the garbage collector in the job, so the job has access to all the properties that are
+     * relevant for its execution, and that do not get passed in via its constructor. Then it retrieves the queue based
+     * on its class and adds it there.
+     *
+     * After adding the job to its corresponding queue we try to consolidate the queue and persist the changes in the
+     * garbage collector state file.
+     *
+     * @param job the job that shall be executed
+     * @throws GarbageCollectorException if anything goes wrong while adding the job
+     */
+    public void addJob(GarbageCollectorJob job) throws GarbageCollectorException {
+        job.registerGarbageCollector(this);
+
+        ArrayDeque<GarbageCollectorJob> jobQueue = getJobQueue(job.getClass());
+        jobQueue.addLast(job);
+
+        // consolidating the queue is optional
+        QueueConsolidator queueConsolidator = queueConsolidators.get(job.getClass());
+        if(queueConsolidator != null) {
+            queueConsolidator.consolidateQueue(this, jobQueue);
+        }
+
+        persistChanges();
+    }
+
+    /**
+     * This method spawns the thread that is taking care of processing the cleanup jobs in the background.
+     *
+     * It repeatedly calls {@link #processCleanupJobs()} until the GarbageCollector is shutting down.
+     */
+    public void start() {
+        (new Thread(() -> {
+            log.info("Snapshot Garbage Collector started ...");
+
+            while(!shuttingDown) {
+                try {
+                    processCleanupJobs();
+
+                    try {
+                        Thread.sleep(GARBAGE_COLLECTOR_RESCAN_INTERVAL);
+                    } catch(InterruptedException e) { /* do nothing */ }
```

Comment:
I know this is how it looks like in the rest of the code.
But actually we are not supposed to ignore interruptedException.
See #605 (that was never merged due to me asking something that wasn't fulfilled), and https://dzone.com/articles/how-to-handle-the-interruptedexception